### PR TITLE
Merge MovesGenerator and Move classes and subclasses, repackage

### PR DESCRIPTION
**Pull request by cherche**
Resolves conflicts between #14, #41, #17.

### CODE REVIEW

File: src/main/java/engine/entities/Board.java [hunk 1]
```diff
@@ -71,16 +100,39 @@ public Move getPreviousMove() {
...
+    }
+
+    /**
+     * Sets the current player to the next one in
+     * the players array. If this is the "last" player,
+     * wrap back around to the first one.
+     * If the current player has yet to be set,
+     * set it to the first player in the array.
+     */
+    public void advanceCurrentPlayer() {
```
Comment (Gursach44) [inline src/main/java/engine/entities/Board.java:129]: We will consider moving this method into the Game class as Board seems not to need the information of which player's turn it is. The process by which turns are executed is outlined here: https://docs.google.com/document/d/1oo0thyl2tfIQTjTtnxEBpouTDv4sWtX3vEecNI4_wAU/edit

This also means we can remove the need for Board to store the current player.
In the case that a new game is started, the game class knows whose turn it is.
In the case that a game is being loaded, moves will be stepped through as if players were enacting them, so this will be no different and the Game class can still keep track of the current turn. 
In the case of undoing moves, we can implement it in such a way that Game still makes it the other player's turn (it is always the case that when a move is being undone, the current player switches anyways, so this should be fine). I think this covers all cases.

BUT this does raise another point; at what stage is the player actually allowed to undo a move? having made a move, it will become the opposite player's turn as we currently have in the outline. Maybe we could ask the player to confirm their move at every turn (i.e. show the board updated with their move made, ask them if they would like to lock it in or undo). Although this raises another issue with the ability to undo multiple moves - if a move is undone, should the player be prompted again, asking whether they'd like to undo again or continue playing? (Sorry if this isn't the best place to discuss this. Also, I guess this issue didn't come up before because we probably would have just had a single undo button that would not rely on the current player turn or anything.)

File: src/main/java/engine/movegen/PawnMovesGen.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        super.removeInvalid(this.moves, board);
+        return this.moves;
+    }
+    /**
+     * Checks to see whether destination of move is already
+     * occupied by any piece. Takes in absolute coordinates.
+     */
+    private boolean isOccupiedByPiece (Move move, Board board) {
+        return isOccupiedByFriendly(move, board) && isOccupiedByEnemy(move, board);
```
Comment (Gursach44) [inline src/main/java/engine/movegen/PawnMovesGen.java]: This is not directly related to the main goal of this PR (sorry) but I'm pretty sure this statement will never return true. I think it was supposed to be an OR rather than an AND.

File: src/main/java/engine/movegen/RookMovesGen.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package engine.movegen;
```
Comment (Gursach44) [inline src/main/java/engine/movegen/RookMovesGen.java:1]: This is an outdated version of the RookMovesGen class, but the QueenMovesGen class in this PR is the updated version. Not sure why this happened. I thought all of the latest MovesGens subclasses were supposed to be in this PR but maybe not? Either way marking this so I don't forget that it has to be changed (I will still forget).

**Review by Gursach44 — DISMISSED** [review]
Almost everything looks great. Definitely some necessary changes but I think they can be made after merging. This is BIG let's MERGE IT woooooooo

File: src/main/java/engine/entities/Board.java (hunk 1 above)
Comment (cherche) [reply src/main/java/engine/entities/Board.java:129]: I remembered why Board is given the player's turn; this is what ensured that it is possible to advance the player's turn from within a `Move` object's `execute()` method. Indeed, it is better to refactor so that most of the "universal" actions associated with executing a move are handled elsewhere. This is the role filled by `MoveTryer` (see CRC cards). Indeed, this can and should be moved out.

I don't quite see the issue with undoing moves. An undo simply goes back to the previous board. Yes, it reverts to the previous player's turn—it is *the previous player's* move which is being undone. This behaviour is consistent with most two-player chess apps.

The only time one would have to undo multiple moves is if they were playing against a bot (in which case there is no phase in which one awaits the bot's move and has an opportunity to undo). Then it makes sense for the undo to revert two boards back to return to the last turn on which it was the non-bot player's turn.

File: src/main/java/engine/movegen/PawnMovesGen.java (hunk 2 above)
Comment (cherche) [reply src/main/java/engine/movegen/PawnMovesGen.java]: Yep, looks like a typo

File: src/main/java/engine/movegen/RookMovesGen.java (hunk 3 above)
Comment (cherche) [reply src/main/java/engine/movegen/RookMovesGen.java:1]: I was under the impression that all the moves generators, including #32, were included and up to date (see discussion in #18).

File: src/main/java/engine/move/PromoteMove.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     * @param board the board on which this move should be executed
+     */
+    @Override
+    public void execute(Board board) {
+        Piece pawn = board.getPieceAtAbsCoords(origin[0], origin[1]);
+        // (No captures)
+        // Move pawn to row
+        movePiece(board, origin, destination);
+        // Promote to new type
+        pawn.setType(promotedType);
```
Comment (cherche) [inline src/main/java/engine/move/PromoteMove.java]: This will not properly work because the point value is not changed. Should write a test to cover this case. An appropriate fix is to create a brand new Piece object instead of trying to mutate the pawn in place.

File: src/main/java/engine/entities/Board.java (hunk 1 above)
Comment (cherche) [reply src/main/java/engine/entities/Board.java:129]: `PostMoveValidator` will need to be based on a Game rather than a Board if we choose this paradigm. Overall, it becomes difficult to work with just Board objects. Much of the program would need to be restructured.

Comment (cherche) [reply src/main/java/engine/entities/Board.java:129]: Actually, this is false. It is still possible to verify whether a move is valid but it must assume that the piece at a move's origin has the colour of the moving player.

File: src/test/java/BoardTest.java [hunk 5]
```diff
@@ -66,4 +69,16 @@ public void BoardMirrorReversesRowsFixesColumns() {
...
+    @Test
+    public void AdvanceCurrentPlayer() {
+        Board board = new Board();
+        Player p1 = new Player("Player 1", 'W');
+        Player p2 = new Player("Player 2", 'B');
+        Player[] players = {p1, p2};
+        board.setPlayers(players);
+        board.setCurrentPlayer(p1);
+        board.advanceCurrentPlayer();
+        Assertions.assertEquals(p2, board.getCurrentPlayer());
```
Comment (adirao-projects) [inline src/test/java/BoardTest.java:82]: Change to `Assertions.assertSame(p2, board.getCurrentPlayer())` as this also checks to see if they are refering to the same object.

File: src/test/java/BoardCreatorTest.java [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
+import engine.entities.*;
```
Comment (adirao-projects) [inline src/test/java/BoardCreatorTest.java:4]: It is better to call the specific classes you need in `engine.entities` rather than calling all of `entities`

File: src/main/java/engine/movegen/RookMovesGen.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                    moveList.add(toAdd);
+                }
+                else {break;}
+            }
+        }
+        //TODO: add ability to castle
+
+        return moveList;
+    }
+}
```
Comment (adirao-projects) [inline src/main/java/engine/movegen/RookMovesGen.java:53]: Add line at end of file (style).

File: src/main/java/engine/movegen/PawnMovesGen.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * Creates list of possible moves a pawn can
+     * make which is later refined by PostMoveValidator.
+     * Checks if pawn can move one square, two squares,
+     * capture diagonally, promote, or en passant and adds
+     * them to the list of moves. Any moves which jump over
+     * a piece, are not on the board, or are occupied are
+     * removed from the list. King checks are NOT considered.
+     */
+    public List<Move> generate (Board board, Piece pawn) {
```
Comment (adirao-projects) [inline src/main/java/engine/movegen/PawnMovesGen.java:19]: Current `generate` method is too complex with too many branches. Consider adding helper functions to increase readability and to assist with future debugging.

File: src/main/java/engine/movegen/MovesGenerator.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    public void removeInvalid (List<Move> moves, Board board) {
+        // TODO: will use checkTile method
+    }
+    /**
+     * Returns true if the destination of move is on
+     * board and is NOT occupied by friendly piece.
+     */
+    public boolean checkTile (Move move, Board board) {
+        // TODO: will use methods containsAbsCoords from board class and isOccupiedByFriendly and
+        return true;
```
Comment (adirao-projects) [inline src/main/java/engine/movegen/MovesGenerator.java:33]: Consider adding the following error to incomplete methods:
`throw new java.lang.UnsupportedOperationException("Not implemented yet.");`

as opposed to simply returning `true`

File: src/main/java/engine/movegen/MovesGenerator.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /**
+     * Checks to see whether destination of move is already
+     * occupied by an enemy piece using absolute coordinates.
+     */
+    public boolean isOccupiedByEnemy (Move move, Board board) {
+        return true;
+    }
+}
```
Comment (adirao-projects) [inline src/main/java/engine/movegen/MovesGenerator.java:50]: Add line at end of file.

**Review by adirao-projects — CHANGES_REQUESTED** [review]

File: src/main/java/engine/movegen/MovesGenerator.java (hunk 9 above)
Comment (cherche) [reply src/main/java/engine/movegen/MovesGenerator.java:33]: We will do from now on whenever writing an incomplete method.

File: src/main/java/engine/movegen/PawnMovesGen.java (hunk 8 above)
Comment (cherche) [reply src/main/java/engine/movegen/PawnMovesGen.java:19]: Please open an issue which says that this method should be refactored and better documented for sake of readability. We will not fix in this PR

**Review by adirao-projects — APPROVED** [review]
All changes look good now

**Review by David9Nagy — APPROVED** [review]

**Review by Gursach44 — APPROVED** [review]
